applications/coreutils/tac.cpp

```diff
@@ -0,0 +1,168 @@
+#include <libsystem/cmdline/CMDLine.h>
+#include <libsystem/io/Path.h>
+#include <libsystem/io/Stream.h>
+#include <libutils/String.h>
+#include <libutils/StringBuilder.h>
+#include <libutils/Vector.h>
+
+#include <cstring>
+
+static bool before = false;
+static char *separator = nullptr;
+
+static const char *usages[] = {
+    "NAME...",
+    "[OPTION]... NAME...",
+    nullptr,
+};
+
+static CommandLineOption options[] = {
+    COMMANDLINE_OPT_HELP,
+    COMMANDLINE_OPT_STRING("separator", 's', separator,
+                           "Choose the separator[STRING] to be used instead of \\n",
+                           COMMANDLINE_NO_CALLBACK),
+    COMMANDLINE_OPT_BOOL("before", 'b', before,
+                         "Attach the separator before instead of after the string",
+                         COMMANDLINE_NO_CALLBACK),
+    COMMANDLINE_OPT_END};
+
+static CommandLine cmdline = CMDLINE(
+    usages,
+    options,
+    "Concatenate and print lines of file(s) in reverse.",
+    "If no filename provided read from input stream\nNote that command may not work as expected when \\0 is encountered");
+
+char *str_split(char *str, char *const sub_str)
+{
+    /*
+    Function to split string into parts with sub_str as delimiter
+    Similar to strtok in string.h
+    May not work as expected when the str contains '\0' in between
+    */
+    static char *start = nullptr;
+
+    if (start == nullptr)
+    {
+        start = str;
+    }
+
+    if (!*start)
+    {
+        return nullptr;
+    }
+
+    char *split = strstr(start, sub_str);
+
+    if (split)
+    {
+        (*split) = 0;
+        char *tmp = start;
+        start = split + strlen(sub_str);
+        return tmp;
+    }
+
+    int len = strlen(start);
+    if (len)
+    {
+        char *tmp = start;
+        start += len;
+        return tmp;
+    }
+
+    return nullptr;
+}
+
+Result tac(Stream *const input_stream)
+{
+    size_t read;
+    char buffer[1024];
+    char *split = nullptr;
+
+    String s;
+    StringBuilder temp;
+    Vector<String> lines(20);
+    if (!separator)
+    {
+        separator = new char[2];
+        strcpy(separator, "\n");
+    }
+
+    while ((read = stream_read(input_stream, buffer, 1024)) != 0)
+    {
+
+        buffer[read] = 0;
+        char *copy = strdup(buffer);
```

Comment:
Why do you need to make a copy of the buffer? Also, you are not freeing this memory.